### DISCUSSION TITLE
Strange clash between Xcode and command-line tools when building Swift system library package

### DISCUSSION THREAD

**Discussion by fwcd** [Homebrew/discussions#4025] [Everyday usage]
I've originally posted this [on the Swift forums](https://forums.swift.org/t/system-library-wrapper-does-not-build-on-macos-13-due-to-module-redefinition-errors/61881), but since it seems to be a Homebrew-related problem with the way `pkg-config` is set up, I'll cross-post this here:

---

Hi everyone, apologies if this is the wrong place to ask.

I'm currently debugging a problem where building a Swift package wrapping a system library errors on macOS 13 (Xcode 14.1, Swift 5.7, with Homebrew installed) with the following message:

```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.sdk/usr/include/ffi/module.modulemap:1:8: error: redefinition of module 'FFI'
module FFI [system] [extern_c] {
       ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.sdk/usr/include/ffi/module.modulemap:1:8: note: previously defined here
module FFI [system] [extern_c] {
```

The issue, from what I understand, is that the compiler tries to include both the Xcode header and the command-line tools header for `libffi`, causing a clash. A little bit of googling lead me to the following threads with similar issues, suggesting it might be an issue with Homebrew:

- https://forums.swift.org/t/redefinition-of-module-issue/35221
- https://forums.swift.org/t/building-with-a-dependency-that-uses-system-libraries-no-longer-works/35249

Still, this feels like a strange issue. I've put together a simple example project that builds fine on Ubuntu, but errors on macOS (the GitHub Actions workflows confirm this):

https://github.com/fwcd/swift-systemlib-demo

The Xcode include seems to be coming from the `Foundation` import in `SystemlibDemo.swift` (removing it fixes the build) and the command-line tools include I am not sure about. Running `pkg-config --cflags libffi` yields

```
-I/Library/Developer/CommandLineTools/SDKs/MacOSX13.sdk/usr/include/ffi
```

which seems to be coming from `/opt/homebrew/Library/Homebrew/os/mac/pkgconfig/13/libffi.pc`:

```
homebrew_sdkroot=/Library/Developer/CommandLineTools/SDKs/MacOSX13.sdk
prefix=${homebrew_sdkroot}/usr
exec_prefix=/usr
libdir=${exec_prefix}/lib
toolexeclibdir=${libdir}
includedir=${prefix}/include/ffi

Name: libffi
Description: Library supporting Foreign Function Interfaces
Version: 3.4-rc1
Libs: -L${toolexeclibdir} -lffi
Cflags: -I${includedir}
```

Any ideas or suggestions on where my configuration went wrong or how to solve this module clash?

**Comment by Bo98** [comment]
Hmm yeah this is a bit tricky.

The core of the issue is the SDK used for the C flag needs to match the SDK used to compile.

For libffi, the include path is not in the default search path, hence the need for a C flag, unlike other pkg-config files which have none. pkg-config has no notion of environment variables within scripts, so the best we could do is a hardcoded default that can be overridden by passing `--define=homebrew_sdkroot=$SDKROOT`.

Ideally what we'd probably do instead is something like `-iwithsysroot /usr/include/ffi`. Unfortunately, this isn't supported by GCC, which is a deal breaker.

Even if we shipped a pkg-config wrapper that passed `--define=homebrew_sdkroot=$SDKROOT` automatically, it wouldn't work with Swift because Swift Package Manager doesn't actually use the `pkg-config` executable - it has its own custom pkg-config parser.

**Reply by carlocab** [reply]
> Ideally what we'd probably do instead is something like `-iwithsysroot /usr/include/ffi`. Unfortunately, this isn't supported by GCC, which is a deal breaker.

Annoyingly, GCC does support something like `-isystem=/usr/include/ffi` for this, but Clang does not.

**Reply by fwcd** [reply]
Hm, so using bottled libraries from Homebrew which are (so I assume) precompiled against the CLI tools with an Xcode toolchain is a problem in general? Shouldn't ABI stability on macOS take care of this?

**Reply by Bo98** [reply]
This is nothing to do with ABI. pkg-config files are purely to help some software find libraries at compile time.

In terms of path stability: only the CLT is at a fixed location. Xcode can be installed absolutely anywhere.

**Comment by fwcd** [comment]
Since my test repo now seems to exhibit this issue [with all Xcode versions back to Xcode 11.7](https://github.com/fwcd/swift-systemlib-demo/actions/runs/3643843243), I think there was some change in Homebrew that changed the way SwiftPM sets its flags.

The problem can be observed in a larger Swift project of mine that extensively depends on packages using system libraries: This change was probably between August 28 ([green CI](https://github.com/fwcd/d2/commit/cb4a5f6db8bbdccc13bba7b26d938c9c65c821c8)) and November 3 ([red CI](https://github.com/fwcd/d2/commit/60a2b884dc74e66af338b3352363c5211794c096)). [No structural changes](https://github.com/fwcd/d2/compare/60a2b884dc74e66af338b3352363c5211794c096..cb4a5f6db8bbdccc13bba7b26d938c9c65c821c8) to my project happened in that period, the builds should be pretty comparable.

**Reply by fwcd** [reply]
Hm, looking at [the PRs in this timespan](https://github.com/Homebrew/brew/pulls?q=is%3Apr+merged%3A2022-08-28..2022-11-03+), the only one that looks relevant is

- https://github.com/Homebrew/brew/pull/14022

but this is about Ruby, so probably unrelated.

**Reply by Bo98** [reply]
The pkg-config files have been unchanged in years so I'm surprised you've seen a change to be honest. I would have expected your issue to have always been an issue since macOS 10.14.

**Comment by fwcd** [comment]
Looks like the issue solved itself with the most recent `brew update` (compare [this run](https://github.com/fwcd/swift-gif/actions/runs/6630731169/job/18012757257) with [this run](https://github.com/fwcd/swift-gif/actions/runs/6630807303/job/18012997684)). Not sure what changed though.

In any case, the [upstream fix](https://github.com/apple/swift-package-manager/pull/6772) in Swift Package Manager should be included in Swift 5.10.
